### PR TITLE
Fix: Manifest generation exhausts memory on Lambda (#1596)

### PR DESCRIPTION
**Pull request by achave11**

### CODE REVIEW

File: src/azul/service/storage_service.py [hunk 1]
```diff
@@ -21,6 +22,7 @@
 AWS_S3_DEFAULT_MINIMUM_PART_SIZE = 5242880  # 5 MB; see https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html
 
 MULTIPART_UPLOAD_MAX_WORKERS = 4
+MULTIPART_UPLOAD_MAX_CONCURRENT_TASKS = 4
```
Comment (hannes-ucsc) [inline src/azul/service/storage_service.py]:
```suggestion
MULTIPART_UPLOAD_MAX_PENDING_PARTS = 4
```

and add a comment explaining what this does. For example, explain what setting this to zero would mean.

File: src/azul/service/storage_service.py [hunk 2]
```diff
@@ -206,9 +210,20 @@ def __abort(self):
         self.thread_pool.shutdown(wait=False)
         logger.warning('Upload %s: Aborted', self.upload_id)
 
+    def _submit(self, fn, *args, **kwargs):
```
Comment (hannes-ucsc) [inline src/azul/service/storage_service.py]: Please give attribution to where we took this idea (and almost verbatim implementation) from. Credit, where credit is due.

File: src/azul/service/storage_service.py [hunk 3]
```diff
@@ -206,9 +210,20 @@ def __abort(self):
...
+        self.semaphore.acquire()
+        try:
+            future = self.thread_pool.submit(fn, *args, **kwargs)
+        except:
+            self.semaphore.release()
+            raise
+        else:
+            future.add_done_callback(lambda x: self.semaphore.release())
```
Comment (hannes-ucsc) [inline src/azul/service/storage_service.py]:
```suggestion
            future.add_done_callback(lambda _future: self.semaphore.release())
```

File: src/azul/service/storage_service.py [hunk 4]
```diff
@@ -21,6 +22,8 @@
 AWS_S3_DEFAULT_MINIMUM_PART_SIZE = 5242880  # 5 MB; see https://docs.aws.amazon.com/AmazonS3/latest/dev/qfacts.html
 
 MULTIPART_UPLOAD_MAX_WORKERS = 4
+MULTIPART_UPLOAD_MAX_PENDING_PARTS = 4  # Sets the amount of pending tasks that can be queued for execution.
+# Setting it to 0 would mean no tasks are allowed to be queued, only running tasks allowed in the thread pool.
```
Comment (danielsotirhos) [inline src/azul/service/storage_service.py]:
```suggestion

# The amount of pending tasks that can be queued for execution. A value of 0
# allows no tasks to be queued, only running tasks allowed in the thread pool.
MULTIPART_UPLOAD_MAX_PENDING_PARTS = 4
```

File: src/azul/service/manifest_service.py [hunk 5]
```diff
@@ -527,7 +527,7 @@ def write_to(self, output: IO[str]) -> Optional[str]:
         writer = csv.DictWriter(output, sources, dialect='excel-tab')
         writer.writeheader()
         project_short_names = set()
-        for hit in self._create_request().scan():
+        for hit in self._create_request().params(size=500).scan():
```
Comment (hannes-ucsc) [inline src/azul/service/manifest_service.py]: Explain the choice of 500 in a comment.

File: src/azul/service/storage_service.py [hunk 6]
```diff
@@ -206,9 +213,21 @@ def __abort(self):
         self.thread_pool.shutdown(wait=False)
         logger.warning('Upload %s: Aborted', self.upload_id)
 
+    def _submit(self, fn, *args, **kwargs):
+        # Method obtained from https://www.bettercodebytes.com/theadpoolexecutor-with-a-bounded-queue-in-python/
```
Comment (hannes-ucsc) [inline src/azul/service/storage_service.py]:
```suggestion
        # Taken from https://www.bettercodebytes.com/theadpoolexecutor-with-a-bounded-queue-in-python/
```

**Pull request merged**